### PR TITLE
Implement sample GAT model for working PyG with DeepChem

### PR DESCRIPTION
**Pull request by nd-02110114**
This PR is a part of #1942 

## What I did

- Create the new molecule graph convolution featurizer using GraphData.
  - The new featurizer return a basic atom features and bond features like GCN, MPNN.
    - The present graph conv features have some values of nodes and edges which the paper like WeaveNet, MEGNet, MPNN didn't mention
   - Create `utils/graph_conv_utils.py`  and the featurizer is more readable and customizable for users
- Create a sample GAT model
  -  This model is an example of working DeepChem with PyTorch Geometric

## TODO

- [x] Add more docstrings
- [x] Add more tests (mainly, all functions in `utils/graph_conv_utils.py`)

I will make another PR. 
- [ ] Debug the problem that my GAT model show the too low score of the overfit tests compared with present TF GraphConvModel
- [ ] Add mode args for GATModel like #2123

### CODE REVIEW

**Comment by nd-02110114** [comment]
This PR is ready to review!

**Comment by rbharath** [comment]
One more quick ask, could you update the model cheatsheet to add `GATModel` and `CGCNNModel`? https://deepchem.readthedocs.io/en/latest/models.html#model-cheatsheet

**Comment by nd-02110114** [comment]
> One more quick ask, could you update the model cheatsheet to add GATModel and CGCNNModel? https://deepchem.readthedocs.io/en/latest/models.html#model-cheatsheet

I will add a lot of documentations related to TorchModel, CGCNN, GAT in #2124

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  construct_hydrogen_bonding_info, get_atom_hydrogen_bonding_one_hot, \
+  get_atom_is_in_aromatic_one_hot, get_atom_hybridization_one_hot, \
+  get_atom_total_num_Hs_one_hot, get_atom_chirality_one_hot, get_atom_formal_charge, \
+  get_atom_partial_charge, get_atom_ring_size_one_hot, get_atom_total_degree_one_hot, \
+  get_bond_type_one_hot, get_bond_is_in_same_ring_one_hot, get_bond_is_conjugated_one_hot, \
+  get_bond_stereo_one_hot
+
+
+def constrcut_atom_feature(atom: RDKitAtom, h_bond_infos: List[Tuple[int, str]],
+                           sssr: List[Sequence]) -> List[float]:
```
[the remark this answers is not in the document]
Comment (nd-02110114) [reply deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:17]: Is there any reasons? I think this function are basically not used in other place. (Currently, this I added the underscore to this function name like `_ construct_atom_feature `)

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  to modify return values of `construct_atom_feature` or `construct_bond_feature`.
+
+  The default node representation are constructed by concatenating the following values,
+  and the feature length is 38.
+
+  - Atom type: A one-hot vector of this atom, "C", "N", "O", "F", "P", "S", "Br", "I", "other atoms".
+  - Chirality: A one-hot vector of the chirality, "R" or "S".
+  - Formal charge: Integer electronic charge.
+  - Partial charge: Calculated partial charge.
+  - Ring sizes: A one-hot vector of the number of rings (3-8) that include this atom.
```
Comment (peastman) [inline deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:88]: I think you mean the size of the ring?  Not many atoms belong to three rings, much less eight!

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  Notes
+  -----
+  This class requires RDKit to be installed.
+  """
+
+  def __init__(self, add_self_loop: bool = False):
+    """
+    Parameters
+    ----------
+    add_self_loop: bool, default False
```
Comment (peastman) [inline deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:132]: Perhaps `add_self_edges` would be clearer?  This isn't really about loops, so the name could be confusing.

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    atom_features = np.array(
+        [
+            _construct_atom_feature(atom, h_bond_infos, sssr)
+            for atom in mol.GetAtoms()
+        ],
+        dtype=np.float,
+    )
+
+    # construct edge (bond) information
+    src, dist, bond_features = [], [], []
```
Comment (peastman) [inline deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:177]: You probably want the variable to be called `dest` (short for destination?), not `dist` which sounds like it's short for "distance"?

File: deepchem/models/torch_models/gat.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+class GATModel(TorchModel):
+  """Graph Attention Networks (GAT).
+
+  Here is a simple example of code that uses the GATModel with
+  molecules dataset.
+
+  >> import deepchem as dc
+  >> featurizer = dc.feat.MolGraphConvFeaturizer()
+  >> dataset_config = {"reload": False, "featurizer": featurizer, "transformers": []}
+  >> tasks, datasets, transformers = dc.molnet.load_tox21(**dataset_config)
```
Comment (peastman) [inline deepchem/models/torch_models/gat.py:132]: Wouldn't it be simpler to just write this as

```
tasks, datasets, transformers = dc.molnet.load_tox21(reload=False, featurizer=featurizer, transformers=[])
```

File: deepchem/models/torch_models/gat.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  Notes
+  -----
+  This class requires PyTorch Geometric to be installed.
+  """
+
+  def __init__(self,
+               in_node_dim: int = 38,
+               hidden_node_dim: int = 64,
+               heads: int = 4,
+               dropout_rate: float = 0.0,
```
Comment (peastman) [inline deepchem/models/torch_models/gat.py:159]: For consistency with other models, this should be just `dropout`.

File: deepchem/models/torch_models/gat.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+               num_conv: int = 3,
+               predicator_hidden_feats: int = 32,
```
Comment (peastman) [inline deepchem/models/torch_models/gat.py:161]: Was that supposed to be "predictor"?

File: deepchem/utils/graph_conv_utils.py [hunk 8]
```diff
@@ -0,0 +1,512 @@
+"""
```
Comment (peastman) [inline deepchem/utils/graph_conv_utils.py:1]: This file doesn't really have anything to do with convolutions.  How about calling it `molecule_feature_utils.py`?

Comment (nd-02110114) [reply deepchem/utils/graph_conv_utils.py]: I fixed!

File: deepchem/models/torch_models/gat.py (hunk 7 above)
Comment (nd-02110114) [reply deepchem/models/torch_models/gat.py:160]: Yes. I fixed!

File: deepchem/models/torch_models/gat.py (hunk 6 above)
Comment (nd-02110114) [reply deepchem/models/torch_models/gat.py]: I fixed!

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py (hunk 4 above)
Comment (nd-02110114) [reply deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py]: This is short for destination. I fixed!

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py (hunk 3 above)
Comment (nd-02110114) [reply deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py]: I agreed! I fixed.

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py (hunk 2 above)
Comment (nd-02110114) [reply deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py]: Yes! I fixed.

File: deepchem/models/torch_models/gat.py (hunk 5 above)
Comment (nd-02110114) [reply deepchem/models/torch_models/gat.py]: I fixed!

File: deepchem/feat/base_classes.py [hunk 9]
```diff
@@ -75,6 +76,60 @@ def _featurize(self, datapoint: Any):
...
+    CGCNNFeaturizer[radius=8.0, max_neighbors=8, step=0.2]
+    """
+    args_spec = inspect.getfullargspec(self.__init__)  # type: ignore
+    args_names = [arg for arg in args_spec.args if arg != 'self']
+    args_info = ''
+    for arg_name in args_names:
+      args_info += arg_name + '=' + str(self.__dict__[arg_name]) + ', '
+    return self.__class__.__name__ + '[' + args_info[:-2] + ']'
+
+  def __str__(self) -> str:
```
Comment (nd-02110114) [inline deepchem/feat/base_classes.py:102]: This implementation is needed to resolve the Windows CI. This is referred #1829. __repr__ function shows all arguments when instantiating a class. __str__ function shows just updated arguments when instantiating a class.

**Comment by nd-02110114** [comment]
This PR is ready to a second review

File: deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  """
+  bond_type = get_bond_type_one_hot(bond)
+  same_ring = get_bond_is_in_same_ring_one_hot(bond)
+  conjugated = get_bond_is_conjugated_one_hot(bond)
+  stereo = get_bond_stereo_one_hot(bond)
+  return bond_type + same_ring + conjugated + stereo
+
+
+class MolGraphConvFeaturizer(MolecularFeaturizer):
+  """This class is a featurizer of gerneral graph convolution networks for molecules.
```
[the remark this answers is not in the document]
Comment (nd-02110114) [reply deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py]: I fixed!

File: deepchem/feat/base_classes.py [hunk 11]
```diff
@@ -75,6 +76,60 @@ def _featurize(self, datapoint: Any):
     """
     raise NotImplementedError('Featurizer is not defined.')
 
+  def __repr__(self) -> str:
```
[the remark this answers is not in the document]
Comment (nd-02110114) [reply deepchem/feat/base_classes.py:79]: > We should add __repr__ methods to all our featurizers :)

I think we don't need to do this. This is because these implementation is put on the `Featurizer` class.

So, this implementation is valid for all featurizers except featurizers inherited ComplexFeaturizer

[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/feat/base_classes.py:79]: Ah you're right!

**Comment by nd-02110114** [comment]
Thanks! I merge in
